lib/configurations: add useHomeManagerModule

yunfachi opened the pull request.

yunfachi committed:
lib/configurations: add useHomeManagerModule

yunfachi committed:
templates: add commented `useHomeManagerModule = false;`

yunfachi committed:
docs/src/hosts: add useHomeManagerModule

akirak reviewed (commented):
Thank you for your quick action! 

It works, but this would mean the user has to make a separate call of `denix.lib.configurations` to disable home-manager on specific hosts.

I wonder it may be simpler to import the home-manager module on the user's side. It would be nice if I were able to define a module as below:

```nix
delib.module {
  name = "home-manager";

  options = delib.singleEnableOption (host.type == "desktop");

  nixos.ifEnabled.imports = [
    home-manager.nixosModules.home-manager
  ];

  darwin.ifEnabled.imports = [
    home-manager.darwinModules.home-manager
  ];

  ...;
}
```

Unfortunately, this isn't possible at present, because `imports` cannot be under `ifEnabled`. 

What do you think?

yunfachi commented on lib/configurations/default.nix:
```diff
@@ -12,8 +12,9 @@
     nixpkgs ? args.nixpkgs,
     home-manager ? args.home-manager,
```
`home-manager`? Not exactly, it's still used when `moduleSystem = "home";` (it's actually optional now, taken from the denix flake)

yunfachi commented:
> I wonder it may be simpler to import the home-manager module on the user's side.

Good idea! I moved `useHomeManagerModule` to the arguments of `delib.module`. I also added `useHomeManagerModule` to `specialArgs` so that it's possible to optionally specify `config.home-manager` (`lib.mkIf host.useHomeManagerModule` can't be used because even when using `lib.mkIf false`, the options still need to be declared):

```nix
{delib, useHomeManagerModule, ...}:
delib.module {
  name = "home-manager";

  nixos.always = if useHomeManagerModule then {home-manager.backupFileExtension="backup";} else {}
  # ...
}
```

akirak commented:
@yunfachi Thank you. Works like a charm.

akirak commented on docs/src/configurations/structure.md:
```diff
@@ -6,6 +6,7 @@
 - `nixpkgs` (nixpkgs): used to override nixpkgs in your configuration, equivalent to `inputs.denix.inputs.nixpkgs.follows`. Defaults to `inputs.nixpkgs`.
 - `home-manager` (home-manager): used to override home-manager in your configuration, equivalent to `inputs.denix.inputs.home-manager.follows`. Defaults to `inputs.home-manager`.
 - `nix-darwin` (nix-darwin): used to override nix-darwin in your configuration, equivalent to `inputs.denix.inputs.nix-darwin.follows`. Defaults to `inputs.nix-darwin`.
+- `useHomeManagerModule` (bool): whether to include the Home Manager module in your NixOS and Nix-Darwin configurations. Defaults to `true`.
```
Maybe `useHomeManager` (or `enableHomeManager`) instead perhaps?

yunfachi replied:
I think `useHomeManager` might not be suitable, as it could be confused with: "do not create this host's configuration at all when `moduleSystem = "home";`"
